core/string/ustring.cpp

```diff
@@ -4545,84 +4545,128 @@ bool String::is_network_share_path() const {
 	return begins_with("//") || begins_with("\\\\");
 }
 
-String String::simplify_path() const {
-	String s = *this;
-	String drive;
-
+static int find_drive_end_idx(const String &p_str) {
 	// Check if we have a special path (like res://) or a protocol identifier.
-	int p = s.find("://");
-	bool found = false;
+	int p = p_str.find("://");
 	if (p > 0) {
 		bool only_chars = true;
 		for (int i = 0; i < p; i++) {
-			if (!is_ascii_alphanumeric_char(s[i])) {
+			if (!is_ascii_alphanumeric_char(p_str[i])) {
 				only_chars = false;
 				break;
 			}
 		}
 		if (only_chars) {
-			found = true;
-			drive = s.substr(0, p + 3);
-			s = s.substr(p + 3);
-		}
-	}
-	if (!found) {
-		if (is_network_share_path()) {
-			// Network path, beginning with // or \\.
-			drive = s.substr(0, 2);
-			s = s.substr(2);
-		} else if (s.begins_with("/") || s.begins_with("\\")) {
-			// Absolute path.
-			drive = s.substr(0, 1);
-			s = s.substr(1);
-		} else {
-			// Windows-style drive path, like C:/ or C:\.
-			p = s.find(":/");
-			if (p == -1) {
-				p = s.find(":\\");
-			}
-			if (p != -1 && p < s.find_char('/')) {
-				drive = s.substr(0, p + 2);
-				s = s.substr(p + 2);
-			}
+			return p + 3;
 		}
 	}
 
-	s = s.replace("\\", "/");
-	while (true) { // in case of using 2 or more slash
-		String compare = s.replace("//", "/");
-		if (s == compare) {
-			break;
-		} else {
-			s = compare;
-		}
+	if (p_str.is_network_share_path()) {
+		// Network path, beginning with // or \\.
+		return 2;
+	}
+
+	if (p_str.begins_with("/") || p_str.begins_with("\\")) {
+		// Absolute path.
+		return 1;
 	}
-	Vector<String> dirs = s.split("/", false);
 
-	for (int i = 0; i < dirs.size(); i++) {
-		String d = dirs[i];
-		if (d == ".") {
-			dirs.remove_at(i);
-			i--;
-		} else if (d == "..") {
-			if (i != 0) {
-				dirs.remove_at(i);
-				dirs.remove_at(i - 1);
-				i -= 2;
+	// Windows-style drive path, like C:/ or C:\.
+	p = p_str.find(":/");
+	if (p == -1) {
+		p = p_str.find(":\\");
+	}
+	if (p != -1) {
+		bool no_separator = true;
+		const char32_t *ptr = p_str.ptr();
+		for (const char32_t *end = ptr + p; ptr < end; ptr++) {
+			if (is_path_seperator(*ptr)) {
+				no_separator = false;
+				break;
 			}
 		}
+		if (no_separator) {
+			return p + 2;
+		}
 	}
 
-	s = "";
+	return 0;
+}
 
-	for (int i = 0; i < dirs.size(); i++) {
-		if (i > 0) {
-			s += "/";
+String String::simplify_path() const {
+	if (is_empty()) {
+		return *this;
+	}
+
+	String ret;
+	// The new string is guaranteed to be equal to or shorter than us.
+	// If shorter, it will be shrunk at the end of the function.
+	ret.resize(size());
+
+	// dst points to the next character to write.
+	char32_t *dst = ret.ptrw();
+
+	// src points to the next character being read.
+	const char32_t *src = ptr();
+	const char32_t *end = src + length();
+
+	// Consume drive part verbatim.
+	for (int drive_end_idx = find_drive_end_idx(*this); drive_end_idx > 0; --drive_end_idx) {
```

Comment:
definitely style, but I would go with : 
`for (int index=0; index< find_drive_end_idx(*this) ; ++index)`

Reply:
It's a bit confusing to read, I confess, but i'm pretty sure not all compilers are able to optimize the additional `int index` away should I introduce it.

Reply:
What do you mean ? You have a variable in both cases. The two cases should compile to something very similar

Reply:
In the `index` example there are two variables (`drive_end_idx` and `index`), in my implementation there is just one (`drive_end_idx`). (besides, for yours there would need to be a line like `int drive_end_idx = find_drive_end_idx(*this) to avoid repeating the call for every iteration).

Reply:
Yeah if you put it above I see your point. Anyway, that's stylistic anyway, and your version is good, so let's resolve this.